pip/pip-278.md

```diff
@@ -0,0 +1,128 @@
+More topic compaction details can be found in [Pulsar Topic Compaction](https://pulsar.apache.org/docs/en/concepts-topic-compaction/).
+
+# Motivation
+
+Currently, the implementation of Pulsar Topic Compaction is fixed and does not support custom strategy, which limits users from using more Compactor policies in their applications.
+
+For example, we need to parse the Kafka format then compact message in Kop, but the current implementation of Pulsar topic compaction does not support this feature.
+
+Another we also need to support write the compacted data anywhere, S3, in columnar format, or even partitioning.
+
+So we need to make  the whole topic compaction service (include Write API & Read API) pluggable to support more customize compaction service implementation.
+
+# Goals
+
+## In Scope
+
+* Abstract topic compaction service interface and support topic compaction service pluggable.
+
+* Migrate the current implementation to a new interface implementation.
+
+* Makes existing tests compatible with new implementations.
+
+## Out of Scope
+
+* For CompactorMetrics, keep the current implementation and don't define related method in topic compaction service interface, in the future will use the `Otel` interface or other mertics API instead.
+
+* For `StrategicTwoPhaseCompactor`, it's out of the scope for regular compaction. It's only used for the load balancer. So it won't change.
+
+
+# High Level Design
+
+To make the whole topic compaction service pluggable, we need to abstract `TopicCompactionService` interface, it can provide the capability that the compactor has and provide the read API to read entries from compacted data.
+
+We should combine `CompactedTopicImpl` and `TwoPhaseCompactor` to pulsar implementation of topic compaction service and make behavior with the current implementation consistent.
+
+
+# Detailed Design
+
+## Design & Implementation Details
+
+* Define a standard TopicCompactionService interface.
+
+  ```java
+  public interface TopicCompactionService {
+      /**
+       * Compact the topic.
+       * Topic Compaction is a key-based retention mechanism. it will keep the most recent value for a given key and
+       * user will read compacted data from TopicCompactionService.
+       *
+       * @return a future that will be completed when the compaction is done.
+       */
+      CompletableFuture<Void> compact();
+
+      /**
+       * Read the compacted entries from the topic.
+       *
+       * @param startPosition         the position to start reading from.
+       * @param numberOfEntriesToRead the number of entries to read.
+       * @return a future that will be completed with the list of entries, this list can is null.
+       */
+      CompletableFuture<List<Entry>> readCompactedEntries(PositionImpl startPosition, int numberOfEntriesToRead);
```

Comment:
Should we consider adding `@NonNull` annotation to interface parameters to clarify it? :)

Reply:
Add `@NonNull` annotation for `List<Entry>`? It seems to be unable to add an annotation for the Generic Types

Reply:
add annotation to `PositionImpl startPosition`

Reply:
Oh, Good idea